### PR TITLE
Bank Register

### PR DESCRIPTION
**Pull request by martinmilev**
[presentation](https://martinmilev.github.io/htmlPages/servlets/bankReg/bank1.html)
Ready for review.

### CODE REVIEW

File: src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  @Before
+  public void setUp() throws Exception {
+    ConnectionProvider provider = new ConnectionProvider("jdbc:mysql://localhost/myBank", "root", "123123");
+    DataStore dataStore = new DataStore(provider);
+    repo = new PersistentAccountRepository(dataStore);
+    dataStore.update("truncate accounts");
+  }
+
+  @Test
+  public void happyPath() throws Exception {
```
Comment (freekman) [inline src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: read the chapter for the unit tests from the book ```Clean code ``` and tell me where are you mistaking if you are mistaking here

File: src/main/java/com/clouway/http/servlets/IndexPageServlet.java [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  protected void doGet(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    getInnerHtml(resp, HttpServletResponse.SC_OK, template.evaluate());
+  }
+
+  @Override
+  protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws ServletException, IOException {
+    register(req, resp);
+  }
+
+  private void register(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/IndexPageServlet.java]: move the logic in the doPost

File: src/main/resources/index.html [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    <meta charset="UTF-8">
+    <title>Home</title>
+    <link rel="stylesheet" href="http://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">
+</head>
+<body>
+<h1 class="text-center">myBank</h1>
+<hr>
+<div class="col-md-2 col-md-offset-5">
+    <form class="form-horizontal" action="/" method='post'>
+        <ul class="nav nav-tabs">
```
Comment (freekman) [inline src/main/resources/index.html]: this ```ul``` is not part of the form

File: src/main/resources/index.html [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+            <li role="presentation" class="active"><a href="#">Register</a></li>
+            <li role="presentation"><a href="#">Login</a></li>
+        </ul>
+        <div class="form-group">
+            <br>${error}
```
Comment (freekman) [inline src/main/resources/index.html]: dont use br

File: src/main/test/com/clouway/http/servlets/IndexPageServletTest.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    String actual = stream.toString();
+    assertThat(response.getStatus(), is(400));
+    assertThat(actual, is(expected));
+  }
+
+  @Test
+  public void register() throws Exception {
+    String name = "John";
+    request.setParameter("name", name);
+    request.setParameter("password", name);
```
Comment (freekman) [inline src/main/test/com/clouway/http/servlets/IndexPageServletTest.java]: .

File: src/main/test/com/clouway/http/servlets/IndexPageServletTest.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+    servlet.doPost(request, response);
+    String actual = stream.toString();
+    assertThat(response.getStatus(), is(400));
+    assertThat(actual, is(expected));
+  }
+
+  @Test
+  public void register() throws Exception {
+    String name = "John";
```
Comment (freekman) [inline src/main/test/com/clouway/http/servlets/IndexPageServletTest.java]: remove this variable it does not make you code more clean

File: src/main/test/com/clouway/http/servlets/IndexPageServletTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    }});
+
+    servlet.doPost(request, response);
+    String actual = stream.toString();
+    assertThat(response.getStatus(), is(400));
+    assertThat(actual, is(expected));
+  }
+
+  @Test
+  public void register() throws Exception {
```
Comment (freekman) [inline src/main/test/com/clouway/http/servlets/IndexPageServletTest.java]: does this test pass

File: src/main/java/com/clouway/http/servlets/IndexPageServlet.java (hunk 2 above)
Comment (martinmilev) [inline src/main/java/com/clouway/http/servlets/IndexPageServlet.java]: fixed

File: src/main/resources/index.html (hunk 3 above)
Comment (martinmilev) [inline src/main/resources/index.html]: fixed

File: src/main/resources/index.html (hunk 4 above)
Comment (martinmilev) [inline src/main/resources/index.html]: fixed

File: src/main/test/com/clouway/http/servlets/IndexPageServletTest.java (hunk 7 above)
Comment (martinmilev) [inline src/main/test/com/clouway/http/servlets/IndexPageServletTest.java]: yes

File: src/main/test/com/clouway/http/servlets/IndexPageServletTest.java (hunk 6 above)
Comment (martinmilev) [inline src/main/test/com/clouway/http/servlets/IndexPageServletTest.java]: fixed

File: src/main/test/com/clouway/http/servlets/IndexPageServletTest.java (hunk 5 above)
Comment (martinmilev) [inline src/main/test/com/clouway/http/servlets/IndexPageServletTest.java]: fixed

**Comment by freekman** [comment]
you have no not done 
```this method is not needed it can be replaced with Optional<A> getByName(...) in the AccountRepo```

File: src/main/java/com/clouway/core/AccountRepository.java [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+import java.util.List;
+
+/**
+ * @author Martin Milev <martinmariusmilev@gmail.com>
+ */
+public interface AccountRepository {
+  void register(Account account);
+
+  boolean exists(String name);
```
Comment (martinmilev) [inline src/main/java/com/clouway/core/AccountRepository.java]: fixed

File: src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    Account account = new Account("John", "123", 123);
+
+    repo.register(account);
+    Account actual = repo.getByName("John").get();
+    Account expected = account;
+
+    assertThat(actual, is(expected));
```
Comment (freekman) [inline src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: why not ```assertThat(actual, is(account));```

File: src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  }
+
+  @Test
+  public void exist() throws Exception {
```
Comment (freekman) [inline src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: 1) bad name for test
2) this test case id tested by the upper test

File: src/main/java/com/clouway/persistent/datastore/DataStore.java [hunk 11]
```diff
@@ -31,13 +32,13 @@ public void update(String query, Object... objects) {
     }
   }
 
-  public <T> List<T> fetchRows(String query, RowFetcher<T> rowFetcher) {
+  public <T> List<T> fetchRows(String query, RowFetcher<Account> rowFetcher) {
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/datastore/DataStore.java]: ?

File: src/main/java/com/clouway/persistent/datastore/RowFetcher.java [hunk 12]
```diff
@@ -1,10 +1,11 @@
...
 
 import java.sql.ResultSet;
+import java.sql.SQLException;
 
 /**
  * @author Martin Milev <martinmariusmilev@gmail.com>
  */
 public interface RowFetcher<T> {
-  T fetchRow(ResultSet resultSet);
+  T fetchRow(ResultSet resultSet) throws SQLException;
```
Comment (freekman) [inline src/main/java/com/clouway/persistent/datastore/RowFetcher.java]: ?

File: src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java (hunk 9 above)
Comment (martinmilev) [inline src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: fixed

File: src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java (hunk 10 above)
Comment (martinmilev) [inline src/main/test/com/clouway/persistent/adapter/jdbc/PersistentAccountRepositoryTest.java]: fixed

File: src/main/java/com/clouway/persistent/datastore/DataStore.java (hunk 11 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/datastore/DataStore.java]: Don't know why this change occurred. fixed

File: src/main/java/com/clouway/persistent/datastore/RowFetcher.java (hunk 12 above)
Comment (martinmilev) [inline src/main/java/com/clouway/persistent/datastore/RowFetcher.java]: .
fixed

File: src/main/java/com/clouway/http/servlets/IndexPageServlet.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+    if (!repository.getByName(name).isPresent()) {
+      String password = req.getParameter("password");
+      Account account = new Account(name, password, 0);
+
+      repository.register(account);
+
+      resp.sendRedirect("/login");
+    } else {
+      template.put("error", "<p style=\"color:red;\">Username is taken.</p>");
```
Comment (freekman) [inline src/main/java/com/clouway/http/servlets/IndexPageServlet.java]: here you should only add the message and not the style

**Comment by freekman** [comment]
LGTM after you fix my last comment
